jerry-core/ecma/builtin-objects/ecma-builtin-date.cpp

```diff
@@ -38,19 +41,258 @@
+
+    if (ecma_number_is_nan (day)
+        || day < 1
+        || day > 31)
+    {
+      goto end; /* NaN */
+    }
+  }
+
+  /* 4. read time if any */
+  if (date_str_size != (lit_utf8_size_t) (date_char_p - date_start_p)
+      && *date_char_p == 'T')
+  {
+    /* 4.1 read hours and minutes */
+    JERRY_ASSERT ((date_char_p - date_start_p) >= 0);
+    uint32_t num_of_visited_chars = (uint32_t) (date_char_p - date_start_p);
+
+    if ((date_str_size - num_of_visited_chars) < 5)
+    {
+      goto end; /* NaN */
+    }
+
+    /* eat up 'T' */
+    date_char_p += sizeof (lit_utf8_byte_t);
+    hours = ecma_date_parse_date_chars (date_start_p, &date_char_p, date_str_size, 2);
+
+    if (ecma_number_is_nan (hours)
+        || hours < 0
+        || hours > 24)
+    {
+      goto end; /* NaN */
+    }
+
+    /* eat up ':' */
+    date_char_p += sizeof (lit_utf8_byte_t);
+
+    minutes = ecma_date_parse_date_chars (date_start_p, &date_char_p, date_str_size, 2);
+
+    if (ecma_number_is_nan (minutes)
+        || minutes < 0
+        || minutes > 59)
+    {
+      goto end; /* NaN */
+    }
+
+    /* 4.2 read seconds if any */
+    if (date_str_size != (lit_utf8_size_t) (date_char_p - date_start_p) && *date_char_p == ':')
+    {
+      /* eat up ':' */
+      date_char_p += sizeof (lit_utf8_byte_t);
+      seconds = ecma_date_parse_date_chars (date_start_p, &date_char_p, date_str_size, 2);
+
+      if (ecma_number_is_nan (seconds)
+          || seconds < 0
+          || seconds > 59)
+      {
+        goto end; /* NaN */
+      }
+
+      /* 4.3 read milliseconds if any */
+      if (date_str_size != (lit_utf8_size_t) (date_char_p - date_start_p) && *date_char_p == '.')
+      {
+        /* eat up '.' */
+        date_char_p += sizeof (lit_utf8_byte_t);
+        milliseconds = ecma_date_parse_date_chars (date_start_p, &date_char_p, date_str_size, 3);
+
+        if (ecma_number_is_nan (milliseconds)
+            || milliseconds < 0)
+        {
+          goto end; /* NaN */
+        }
+      }
+
+      time = ecma_date_make_time (hours, minutes, seconds, milliseconds);
+    }
+
+    /* 4.4 read timezone if any */
+    if (date_str_size != (lit_utf8_size_t) (date_char_p - date_start_p) && *date_char_p == 'Z')
+    {
+      date_char_p += sizeof (lit_utf8_byte_t);
+      time = ecma_date_utc (ecma_date_make_time (hours,
+                                                 minutes,
+                                                 seconds,
+                                                 milliseconds));
+    }
+    else if (date_str_size != (lit_utf8_size_t) (date_char_p - date_start_p)
+             && (*date_char_p == '+' || *date_char_p == '-'))
+    {
+      JERRY_ASSERT ((date_char_p - date_start_p) >= 0);
+      uint32_t num_of_visited_chars = (uint32_t) (date_char_p - date_start_p);
+
+      if ((date_str_size - num_of_visited_chars) < 6)
+      {
+        goto end; /* NaN */
```

Comment:
@LaszloLango, yes, we should avoid using of `goto` everywhere.

Unfortunately, there too many style issues in `parser.cpp` to use it as a reference.
The style issues in parser, as in other engine parts, definitely, should be fixed.
Style issues in parser were fixed step by step, in the context of recent development of feature support, and, probably, style fixing would be continued in context of parser optimization. I think, if after the changes there still would be any styles issues, then the issues should be fixed with a separate style fix patch.